### DISCUSSION TITLE
Supports to merge selected commit to current branch

### DISCUSSION THREAD

**Discussion by wl2776** [sourcegit-scm/sourcegit#800] [Ideas]
In the image below, I've made right-click on the commit, highlighted with blue.

I think, it would be convenient to have an option for merging/fast-forwarding of the checked out branch to the selected commit.
This option is already available from the left panel with branches, adding it here saves several mouse clicks.

![merge-suggestion](https://github.com/user-attachments/assets/fc4dfc98-6c86-4783-8721-8bdd25450d8a)

**Comment by love-linger** [comment]
Currently, we only supports merge branch/tag.

![image](https://github.com/user-attachments/assets/8696857f-05fa-40c4-a651-b74228e16508)

![image](https://github.com/user-attachments/assets/e65f878a-3460-481b-8a9c-b741a96615de)

**Reply by wl2776** [reply]
Options on your are different from my suggestion. Your screenshots show how to merge some branch or tag *into checked out branch*.

I suggest to add option for *opposite direction*: merging of *checked out branch into* some other branch or tag.

**Reply by love-linger** [reply]
It's not a `git merge` command.

![image](https://github.com/user-attachments/assets/fb042fb0-2a4f-45ea-a944-570ee39c631e)

The `git merge` command always merges one or more branches into the **CURRENT** branch.

To do the job shows in your picture, it needs several steps:

* Remember the current branch, let's just call it `ORG_BRANCH`
* Checkout selected commit with a new branch
* Merge the `ORG_BRANCH` into the created new branch
* Resolve conflicts if it is needed
* Cherry-pick all the commits after selected in order.
* Resolve conflicts if it is needed
* After that the created new branch is what you want.

**Reply by love-linger** [reply]
BTW. Imagine this situation, if we do NOT checkout the target branch, and there are some conflicts while merging, which branch is `--mine` (or HEAD changes) and which branch is `--theirs` (or INCOMING changes)? Will these files exist in the worktree which they are not belongs to (since we do NOT checkout that branch)?

**Reply by wl2776** [reply]
> It's not a `git merge` command.
> The `git merge` command always merges one or more branches into the **CURRENT** branch.
> 
> To do the job shows in your picture, it needs several steps:
> 
> * Remember the current branch, let's just call it `ORG_BRANCH`
> * Checkout selected commit with a new branch
> * Merge the `ORG_BRANCH` into the created new branch
> * Resolve conflicts if it is needed
> * Cherry-pick all the commits after selected in order.
> * Resolve conflicts if it is needed
> * After that the created new branch is what you want.

I'm afraid, you're mistaken.

Consider the following setup (in the sourcegit's repo)
![image](https://github.com/user-attachments/assets/cf86ba84-a656-4795-9e57-50b4e1533155)

I've made a new branch `feature/test` at c8cdda3477 and want to incorporate changes from 9b88db29:
Here is how I do it from command line.

Make sure that we are at desired commit

```
$ git checkout feature/test
Switched to branch 'feature/test'

$ git status
On branch feature/test
nothing to commit, working tree clean
```

Then I merge:

```
$ git merge --no-commit --ff 9b88db29
Updating c8cdda34..9b88db29
Fast-forward
README.md                            |  2 +-
TRANSLATION.md                       | 24 ++++++++++++++++++------
src/Resources/Locales/en_US.axaml    |  2 ++
src/Resources/Locales/zh_CN.axaml    |  2 ++
src/Resources/Locales/zh_TW.axaml    |  2 ++
src/ViewModels/InProgressContexts.cs | 84 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++-------------------------
src/ViewModels/WorkingCopy.cs        | 43 +++++++++++++++++++++++++++++++++++++++++++
src/Views/Repository.axaml           |  6 +++++-
8 files changed, 132 insertions(+), 33 deletions(-)
```

Sourcegit reflects the changes

![image](https://github.com/user-attachments/assets/49b73106-ebc5-46de-bda8-00690a189de2)

**Reply by wl2776** [reply]
If I omit `--ff` the result is the same, since 9b88db29 is direct ancestor of c8cdda3

**Reply by wl2776** [reply]
Next, I've made a couple of commits in `feature/test` and repeat merge

Before:
![image](https://github.com/user-attachments/assets/3fd04bb6-6542-4f24-988d-3d5c623a5a28)

```
$ git merge --ff 9b88db29   
Merge made by the 'ort' strategy.
README.md                            |  2 +-
TRANSLATION.md                       | 24 ++++++++++++++++++------
src/Resources/Locales/en_US.axaml    |  2 ++
src/Resources/Locales/zh_CN.axaml    |  2 ++
src/Resources/Locales/zh_TW.axaml    |  2 ++
src/ViewModels/InProgressContexts.cs | 84 +++++++++++++++++++++++++++++++++++++++++++++++++++++++++++-------------------------
src/ViewModels/WorkingCopy.cs        | 43 +++++++++++++++++++++++++++++++++++++++++++
src/Views/Repository.axaml           |  6 +++++-
8 files changed, 132 insertions(+), 33 deletions(-)

```

After: 
![image](https://github.com/user-attachments/assets/3ac6545d-1920-43be-a440-7d0cc6e6dab5)

Please, note that I've left `--ff`, but it doesn't have any effect.

**Reply by wl2776** [reply]
Ah! But it is not "Merge current branch *to* here".  My actions reflect scenario "Merge *selected commit* into current branch"

**Reply by love-linger** [reply]
If you mean `merge selected commit into current branch`, as I said above: 

> Currently, we only supports merge branch/tag.

**Reply by wl2776** [reply]
> If you mean `merge selected commit into current branch`, as I said above:
> 
> > Currently, we only supports merge branch/tag.

I see. It would be nice to have such option for arbitrary commits.

**Reply by wl2776** [reply]
Is it right that this feature should be added in the method `MakeContextMenu` in file Histories.cs?

**Reply by love-linger** [reply]
Done. You can download the latest CI build from Github Action.
